format/udf-spec.md

```diff
@@ -0,0 +1,290 @@
+---
+title: "SQL UDF Spec"
+---
+<!--
+ - Licensed to the Apache Software Foundation (ASF) under one or more
+ - contributor license agreements.  See the NOTICE file distributed with
+ - this work for additional information regarding copyright ownership.
+ - The ASF licenses this file to You under the Apache License, Version 2.0
+ - (the "License"); you may not use this file except in compliance with
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Iceberg SQL UDF Spec
+
+## Background and Motivation
+
+A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters, executes a function body.
+Depending on the function type, the result can be:
+
+- **UDFs** – return a single value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).
+- **Table functions (UDTFs)** – return a table, i.e., a table with zero or more rows and columns.
+
+Many compute engines (e.g., Spark, Trino) already support UDFs, but in different and incompatible ways. Without a common
+standard, UDFs cannot be reliably shared across engines or reused in multi-engine environments.
+
+This specification introduces a standardized metadata format for UDFs in Iceberg. 
+
+## Goals
+
+* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs. 
+* Support function evolution through versioning and rollback.
+* Provide consistent semantics for representing UDFs across engines.
+
+## Overview
+
+UDF metadata follows the same design principles as Iceberg table and view metadata: each function is represented by a
+**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
+properties, and engine-specific representations.
+
+* Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file. The old file is replaced atomically.
+* Each metadata file includes recent definition versions, enabling rollbacks without external state.
+
+## Specification
+
+### UDF Metadata
+The UDF metadata file has the following fields:
+
+| Requirement | Field name                   | Description                                                      |
```

Comment:
I feel like we are missing the type information for each of the defined fields.  Are they a primitive (like a string/Boolean) or a structured/nested type defined below.

Reply:
Added type for each field, please take a look.